lib/integrations/awsoidc/deployservice.go

```diff
@@ -0,0 +1,696 @@
+
+		cluster := describeClustersResponse.Clusters[0]
+		if cluster.Status == nil {
+			return retryutils.PermanentRetryError(trace.Errorf("cluster %q has an unknown (nil) status", *cluster.ClusterName))
+		}
+
+		if *cluster.Status == clusterStatusActive {
+			return nil
+		}
+
+		if *cluster.Status == clusterStatusProvisioning {
+			return trace.Errorf("cluster %q is provisioning...", *cluster.ClusterName)
+		}
+
+		return retryutils.PermanentRetryError(trace.Errorf("unexpected status %s for ECS Cluster %q", *cluster.ClusterName, *cluster.Status))
+	})
+
+	return trace.Wrap(err)
+}
+
+// upsertService creates or updates the service.
+// If the service exists but its LaunchType is not Fargate, then it gets re-created.
+func upsertService(ctx context.Context, clt DeployServiceClient, req DeployServiceRequest, taskARN string) (*ecsTypes.Service, error) {
+	describeServiceOut, err := clt.DescribeServices(ctx, &ecs.DescribeServicesInput{
+		Services: []string{*req.ServiceName},
+		Cluster:  req.ClusterName,
+		Include: []ecsTypes.ServiceField{
+			ecsTypes.ServiceFieldTags,
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Service already exists.
+	if len(describeServiceOut.Services) > 0 {
+		service := &describeServiceOut.Services[0]
+
+		if service.Status == nil {
+			return nil, trace.Errorf("unknown status for ECS Service %q", *req.ServiceName)
+		}
+
+		if *service.Status == serviceStatusDraining {
+			return nil, trace.Errorf("ECS Service is draining, please retry in a couple of minutes")
+		}
+
+		if *service.Status == serviceStatusActive {
+			ownershipTags := req.ResourceCreationTags
+			if !ownershipTags.MatchesECSTags(service.Tags) {
+				return nil, trace.Errorf("ECS Service %q already exists but is not managed by Teleport. "+
+					"Add the following tags to allow Teleport to manage this service: %s", *req.ServiceName, req.ResourceCreationTags)
+			}
+
+			// If the LaunchType is the required one, than we can update the current Service.
+			// Otherwise we have to delete it.
+			if service.LaunchType == ecsTypes.LaunchTypeFargate {
+				updateServiceResp, err := clt.UpdateService(ctx, &ecs.UpdateServiceInput{
+					Service:        req.ServiceName,
+					DesiredCount:   &oneAgent,
+					TaskDefinition: &taskARN,
+					Cluster:        req.ClusterName,
+					NetworkConfiguration: &ecsTypes.NetworkConfiguration{
+						AwsvpcConfiguration: &ecsTypes.AwsVpcConfiguration{
+							AssignPublicIp: ecsTypes.AssignPublicIpEnabled, // no internet connection otherwise
+							Subnets:        req.SubnetIDs,
+						},
+					},
+					ForceNewDeployment: true,
+					PropagateTags:      ecsTypes.PropagateTagsService,
+				})
+				if err != nil {
+					return nil, trace.Wrap(err)
+				}
+
+				return updateServiceResp.Service, nil
+			}
+		}
+
+		// The service is deleted if the its status is INACTIVE (aka deleted) or if it's owned by Teleport and has the wrong LaunchType.
+		_, err := clt.DeleteService(ctx, &ecs.DeleteServiceInput{
```

Comment:
I'm a little concerned about just deleting stuff in the customer's infra. What is this needed for?

Reply:
Why would it have invalid launch type if Teleport created it? Maybe I'm missing something.

Reply:
It shouldn't, and will only happen if something else changed it (some other integration, user did it manually using aws console, ...).

I'm just being cautious here.
This is trying to recover the system from a bad state.

Should I remove it? On the plus side, it's one less permission that we require 😅

Reply:
Yeah in that case I'd remove it for now tbh.

Reply:
Removed :+1: